README.rst
===============
LexicalRichness
===============
.. image:: https://badge.fury.io/py/lexicalrichness.svg
        :target: https://pypi.org/project/lexicalrichness/
.. image:: https://img.shields.io/conda/vn/conda-forge/lexicalrichness.svg
        :target: https://anaconda.org/conda-forge/lexicalrichness
.. image:: https://img.shields.io/conda/pn/conda-forge/lexicalrichness   
	:target: https://anaconda.org/conda-forge/lexicalrichness
.. image:: https://badgen.net/github/release/Naereen/Strapdown.js
        :target: https://github.com/LSYS/LexicalRichness.js/releases

.. image:: https://github.com/LSYS/LexicalRichness/blob/housekeep/images/ghdocbadge.svg
        :target: https://github.com/LSYS/LexicalRichness/blob/master/README.rst
	
.. image:: https://www.codefactor.io/repository/github/lsys/lexicalrichness/badge
        :target: https://www.codefactor.io/repository/github/lsys/lexicalrichness  
.. image:: https://img.shields.io/lgtm/grade/python/g/LSYS/LexicalRichness.svg?logo=lgtm&logoWidth=18)
        :target: https://lgtm.com/projects/g/LSYS/LexicalRichness/context:python

.. image:: https://img.shields.io/pypi/pyversions/lexicalrichness   
	:target: https://img.shields.io/pypi/pyversions/lexicalrichness  
.. image:: https://img.shields.io/badge/Maintained%3F-yes-green.svg
   :target: https://GitHub.com/Naereen/StrapDown.js/graphs/commit-activity

.. |Maintenance yes| image:: https://img.shields.io/badge/Maintained%3F-yes-green.svg
   :target: https://GitHub.com/Naereen/StrapDown.js/graphs/commit-activity
	

.. image:: https://img.shields.io/badge/PRs-welcome-brightgreen.svg
        :target: http://makeapullrequest.com
.. image:: https://img.shields.io/badge/License-MIT-blue.svg
        :target: https://lbesson.mit-license.org
.. image:: https://mybinder.org/badge_logo.svg
        :target: https://mybinder.org/v2/gh/LSYS/lexicaldiversity-example/main?labpath=example.ipynb
	
	
A small python module to compute textual lexical richness (aka lexical diversity) measures.

Lexical richness refers to the range and variety of vocabulary deployed in a text by a speaker/writer (McCarthy and Jarvis 2007). Lexical richness is used interchangeably with lexical diversity, lexical variation, lexical density, and vocabulary richness and is measured by a wide variety of indices. Uses include (but not limited to) measuring writing quality, vocabulary knowledge (Šišková 2012), speaker competence, and socioeconomic status (McCarthy and Jarvis 2007).

.. image:: http://ForTheBadge.com/images/badges/made-with-python.svg
   :target: https://www.python.org/

Installation
------------
**Install using PIP**

.. code-block:: bash

	pip install lexicalrichness

If you encounter, 

.. code-block:: python

	ModuleNotFoundError: No module named 'textblob'

install textblob:

.. code-block:: bash

	pip install textblob

*Note*: This error should only exist for :code:`versions <= v0.1.3`. Fixed in 
`v0.1.4 <https://github.com/LSYS/LexicalRichness/releases/tag/0.1.4>`__ by `David Lesieur <https://github.com/davidlesieur>`__ and `Christophe Bedetti <https://github.com/cbedetti>`__.


**Install from Conda-Forge**

*LexicalRichness* is now also available on conda-forge. If you have are using the `Anaconda <https://www.anaconda.com/distribution/#download-section>`__ or `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`__ distribution, you can create a conda environment and install the package from conda.

.. code-block:: bash

	conda create -n lex
	conda activate lex 
	conda install -c conda-forge lexicalrichness

*Note*: If you get the error :code:`CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'` with :code:`conda activate lex` in *Bash* either try

	* :code:`conda activate bash` in the *Anaconda Prompt* and then retry :code:`conda activate lex` in *Bash*
	* or just try :code:`source activate lex` in *Bash*

**Install manually using Git and GitHub**

.. code-block:: bash

	git clone https://github.com/LSYS/LexicalRichness.git
	cd LexicalRichness
	pip install .

**Run from the cloud**

Try the package on the cloud (without setting anything up on your local machine) by clicking the icon here:  

|mybinder|

.. |mybinder| image:: https://mybinder.org/badge_logo.svg
 :target: https://mybinder.org/v2/gh/LSYS/lexicaldiversity-example/main?labpath=example.ipynb

Quickstart
----------

.. code-block:: python

	>>> from lexicalrichness import LexicalRichness

	# text example
	>>> text = """Measure of textual lexical diversity, computed as the mean length of sequential words in
            		a text that maintains a minimum threshold TTR score.

            		Iterates over words until TTR scores falls below a threshold, then increase factor
            		counter by 1 and start over. McCarthy and Jarvis (2010, pg. 385) recommends a factor
            		threshold in the range of [0.660, 0.750].
            		(McCarthy 2005, McCarthy and Jarvis 2010)"""

	# instantiate new text object (use the tokenizer=blobber argument to use the textblob tokenizer)
	>>> lex = LexicalRichness(text)

	# Return word count.
	>>> lex.words
	57

	# Return (unique) word count.
	>>> lex.terms
	39

	# Return type-token ratio (TTR) of text.
	>>> lex.ttr
	0.6842105263157895

	# Return root type-token ratio (RTTR) of text.
	>>> lex.rttr
	5.165676192553671

	# Return corrected type-token ratio (CTTR) of text.
	>>> lex.cttr
	3.6526846651686067

	# Return mean segmental type-token ratio (MSTTR).
	>>> lex.msttr(segment_window=25)
	0.88

	# Return moving average type-token ratio (MATTR).
	>>> lex.mattr(window_size=25)
	0.8351515151515151

	# Return Measure of Textual Lexical Diversity (MTLD).
	>>> lex.mtld(threshold=0.72)
	46.79226361031519

	# Return hypergeometric distribution diversity (HD-D) measure.
	>>> lex.hdd(draws=42)
	0.7468703323966486

	# Return Herdan's lexical diversity measure.
	>>> lex.Herdan
	0.9061378160786574

	# Return Summer's lexical diversity measure.
	>>> lex.Summer
	0.9294460323356605

	# Return Dugast's lexical diversity measure.
	>>> lex.Dugast
	43.074336212149774

	# Return Maas's lexical diversity measure.
	>>> lex.Maas
	0.023215679867353005

Attributes and properties
+++++++++++++++++++++++++

+-------------------------+-----------------------------------------------------------------------------------+
| ``wordlist``            | list of words                                                   		      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``words``  		  | number of words (w) 				   			      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``terms``		  | number of unique terms (t)			                                      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``preprocessor``        | preprocessor used		                                                      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``tokenizer``           | tokenizer used		                                                      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``ttr``		  | type-token ratio computed as t / w (Chotlos 1944, Templin 1957)         	      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``rttr``	          | root TTR computed as t / sqrt(w) (Guiraud 1954, 1960)                             |
+-------------------------+-----------------------------------------------------------------------------------+
| ``cttr``	          | corrected TTR computed as t / sqrt(2w) (Carrol 1964)		              |
+-------------------------+-----------------------------------------------------------------------------------+
| ``Herdan`` 	          | log(t) / log(w) (Herdan 1960, 1964)                                               |
+-------------------------+-----------------------------------------------------------------------------------+
| ``Summer``    	  | log(log(t)) / log(log(w)) Summer (1966)                                           |
+-------------------------+-----------------------------------------------------------------------------------+
| ``Dugast``          	  | (log(w) ** 2) / (log(w) - log(t) Dugast (1978)				      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``Maas`` 	          | (log(w) - log(t)) / (log(w) ** 2) Maas (1972)                                     |
+-------------------------+-----------------------------------------------------------------------------------+

Methods
+++++++

+-------------------------+-----------------------------------------------------------------------------------+
| ``msttr``            	  | Mean segmental TTR (Johnson 1944)						      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``mattr``  		  | Moving average TTR (Covington 2007, Covington and McFall 2010)		      |
+-------------------------+-----------------------------------------------------------------------------------+
| ``mtld``		  | Measure of Lexical Diversity (McCarthy 2005, McCarthy and Jarvis 2010)            |
+-------------------------+-----------------------------------------------------------------------------------+
| ``hdd``                 | HD-D (McCarthy and Jarvis 2007)                                                   |
+-------------------------+-----------------------------------------------------------------------------------+

Assessing method docstrings
---------------------------
.. code-block:: python

	>>> import inspect

	# docstring for hdd (HD-D)
	>>> print(inspect.getdoc(LexicalRichness.hdd))

	Hypergeometric distribution diversity (HD-D) score.

	For each term (t) in the text, compute the probabiltiy (p) of getting at least one appearance
	of t with a random draw of size n < N (text size). The contribution of t to the final HD-D
	score is p * (1/n). The final HD-D score thus sums over p * (1/n) with p computed for
	each term t. Described in McCarthy and Javis 2007, p.g. 465-466.
	(McCarthy and Jarvis 2007)

	Parameters
	__________
	draws: int
	    Number of random draws in the hypergeometric distribution (default=42).

	Returns
	_______
	float
	
Alternatively, just do

.. code-block:: python

	>>> print(lex.hdd.__doc__)
	
	Hypergeometric distribution diversity (HD-D) score.

            For each term (t) in the text, compute the probabiltiy (p) of getting at least one appearance
            of t with a random draw of size n < N (text size). The contribution of t to the final HD-D
            score is p * (1/n). The final HD-D score thus sums over p * (1/n) with p computed for
            each term t. Described in McCarthy and Javis 2007, p.g. 465-466.
            (McCarthy and Jarvis 2007)

            Parameters
            ----------
            draws: int
                Number of random draws in the hypergeometric distribution (default=42).

            Returns
            -------
            float	

Contributing
------------
Author
++++++
Lucas Shen

Contributors
++++++++++++
* `Christophe Bedetti <https://github.com/cbedetti>`__
* `David Lesieur <https://github.com/davidlesieur>`__

Contributions are welcome, and they are greatly appreciated! Every little bit helps, and credit will always be given. 
See here for `how to contribute  <./CONTRIBUTING.rst>`__ to this project.
See here for `Contributor Code of
Conduct <http://contributor-covenant.org/version/1/0/0/>`__.

License
-------

The package is released under the `MIT
License <https://opensource.org/licenses/MIT>`__.


HISTORY.rst
=======
History
=======

0.1.2 (2018-05-09)
------------------

* First release on PyPI.

0.1.3 (2018-05-27)
------------------

* Minor fix for compatibility issue with hyphens (ascii) in python 2.

0.1.4 (2021-11-13)
------------------
* Add textblob in setup.py as requirements (to fix "ModuleNotFoundError: No module named 'textblob'") [Christophe Bedetti].
* Make preprocessing and tokenization optional [David Lesieur].

tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for lexicalrichness."""


tests/test_lexicalrichness.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `lexicalrichness` package."""


import unittest

from lexicalrichness import *
# import textblob


class TestLexicalrichness(unittest.TestCase):
    """Tests for `lexicalrichness` package."""

    def setUp(self):
        self.s1 = "TEST text with some text numbers 42, hyphen-here, and text punctuations."
        self.s2 = "TEST text with some text numbers 42, hyphen–here, and text punctuations."
        self.s3 = "TEST text with some text numbers 42, hyphen—here, and text punctuations."
        self.wordlist = "TEST text with some text numbers hyphenhere and text punctuations".split()
        self.emptystring = ''

        self.obj1 = LexicalRichness(self.s1)
        self.obj2 = LexicalRichness(self.s2)
        self.obj3 = LexicalRichness(self.s3)
        self.emptyobj = LexicalRichness(self.emptystring)
        self.wordlistobj = LexicalRichness(self.wordlist, preprocessor=None, tokenizer=None)

    def tearDown(self):
        pass

    def test_preprocess(self):
        self.assertEqual(preprocess(self.s1), 'test text with some text numbers , hyphenhere, and text punctuations.')
        self.assertEqual(preprocess(self.s2), 'test text with some text numbers , hyphenhere, and text punctuations.')
        self.assertEqual(preprocess(self.s3), 'test text with some text numbers , hyphenhere, and text punctuations.')
        self.assertEqual(preprocess(self.emptystring), '')

    def test_tokenize(self):
        self.assertIs(type(tokenize(self.s1)), list)
        self.assertIs(type(tokenize(self.s2)), list)
        self.assertIs(type(tokenize(self.s3)), list)
        self.assertIs(type(tokenize(self.emptystring)), list)

    # def test_blobber(self):
    #     self.assertIs(type(blobber(self.s1)), textblob.blob.WordList)
    #     self.assertIs(type(blobber(self.s2)), textblob.blob.WordList)
    #     self.assertIs(type(blobber(self.s3)), textblob.blob.WordList)
    #     self.assertIs(type(blobber(self.emptystring)), textblob.blob.WordList)

    def test_list_sliding_window(self):
        test_list = ['a', 'b', 'c', 'd']

        self.assertEqual(list(list_sliding_window(test_list, 1)), [('a',), ('b',), ('c',), ('d',)])
        self.assertEqual(list(list_sliding_window(test_list, 2)), [('a', 'b'), ('b', 'c'), ('c', 'd')])
        self.assertEqual(list(list_sliding_window(test_list, 3)), [('a', 'b', 'c'), ('b', 'c', 'd')])
        self.assertEqual(list(list_sliding_window(test_list, 4)), [('a', 'b', 'c', 'd')])

    def test_segment_generator(self):
        test_list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

        self.assertEqual(list(segment_generator(test_list, 3)), [[1,2,3], [4,5,6], [7,8,9], [10]])
        self.assertEqual(list(segment_generator(test_list, 5)), [[1,2,3,4,5], [6,7,8,9,10]])
        self.assertEqual(list(segment_generator(test_list, 1)), [[1], [2], [3], [4], [5], [6], [7], [8], [9], [10]])
        self.assertEqual(list(segment_generator(test_list, 10)), [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]])
        self.assertEqual(list(segment_generator(test_list, 11)), [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]])

    def test_words(self):
        self.assertEqual(self.obj1.words, 10)
        self.assertEqual(self.wordlistobj.words, 10)
        self.assertEqual(self.emptyobj.words, 0)

    def test_terms(self):
        self.assertEqual(self.obj1.terms, 8)
        self.assertEqual(self.wordlistobj.terms, 8)
        self.assertEqual(self.emptyobj.terms, 0)

    def test_ttr(self):
        self.assertEqual(self.obj1.ttr, 0.8)

    def test_rttr(self):
        self.assertEqual(self.obj1.rttr, 2.5298221281347035)

    def test_cttr(self):
        self.assertEqual(self.obj1.cttr, 1.7888543819998317)

    def test_Herdan(self):
        self.assertEqual(self.obj1.Herdan, 0.9030899869919434)

    def test_Summer(self):
        self.assertEqual(self.obj1.Summer, 0.8777828395738175)

    def test_Dugast(self):
        self.assertEqual(self.obj1.Dugast, 23.760032854423635)

    def test_Maas(self):
        self.assertEqual(self.obj1.Maas, 0.04208748389057132)

    def test_msttr(self):
        self.assertEqual(self.obj1.msttr(segment_window=5, discard=True), 0.8)
        self.assertEqual(self.obj1.msttr(segment_window=5, discard=False), 0.9)

        with self.assertRaises(ValueError):
            self.obj1.msttr(segment_window=0)
        with self.assertRaises(ValueError):
            self.obj1.msttr(segment_window=-1)
        with self.assertRaises(ValueError):
            self.obj1.msttr(segment_window=1.5)

    def test_mattr(self):
        self.assertEqual(self.obj1.mattr(window_size=5), 0.9)
        self.assertEqual(self.obj1.mattr(window_size=1), 1)
        self.assertEqual(self.obj1.mattr(window_size=self.obj1.words), self.obj1.ttr)

        with self.assertRaises(ValueError):
            self.obj1.mattr(window_size=0)
        with self.assertRaises(ValueError):
            self.obj1.mattr(window_size=-1)
        with self.assertRaises(ValueError):
            self.obj1.mattr(window_size=1.5)

    def test_mtld(self):
        self.assertEqual(self.obj1.mtld(threshold=0.72), 14.000000000000004)

        all_unqiue = LexicalRichness('only unique terms in this little string')
        self.assertEqual(all_unqiue.mtld(threshold=0.72), all_unqiue.words)

    def test_hdd(self):
        self.assertEqual(self.obj1.hdd(draws=5), 0.8833333333333332)

        with self.assertRaises(ValueError):
            self.obj1.hdd(draws=0)
        with self.assertRaises(ValueError):
            self.obj1.hdd(draws=-5)
        with self.assertRaises(ValueError):
            self.obj1.hdd(draws=1.5)


if __name__ == '__main__':
    unittest.main()


lexicalrichness/lexicalrichness.py
#  -*-  coding:  utf-8  -*-

from __future__ import division

import sys
if sys.version_info[0] == 3:
    from statistics import mean

from collections import Counter
from itertools import islice
import re
import string
from math import sqrt, log
from scipy.stats import hypergeom

try:
    from textblob import TextBlob
except ImportError:
    pass
else:
    def blobber(text):
        """ Tokenize text into a list of tokens using TextBlob.

            Parameter
            ---------
            text: string

            Return
            ------
            TextBlob list of words
        """
        blob = TextBlob(text)
        return blob.words


def preprocess(text):
    """ 1. lower case
        2. removes digits
        3. removes variations of dashes and hyphens

        Parameter
        ---------
        text: string

        Returns
        -------
        string
    """
    # lowercase and remove digits
    text = re.sub(r'[0-9]+', '', text.lower())

    # Replace dashes/hyphens
    if sys.version_info[0] == 3:
        text = text.replace('–', '')
        text = text.replace('—', '')
        text = text.replace('-', '')
    else:
        text = text.replace('-', '')
    return text


def tokenize(text):
    """ Tokenize text into a list of tokens using built-in methods.

        Parameter
        ---------
        text: string

        Returns
        -------
        list
        """
    text = preprocess(text)

    for p in list(string.punctuation):
        text = text.replace(p, ' ')

    words = text.split()
    return words


def segment_generator(List, segment_size):
    """ Split a list into s segments of size r (segment_size).

        Parameters
        ----------
        List: list
            List of items to be segmented.
        segment_size: int
            Size of each segment.

        Returns
        -------
        Generator
    """
    for i in range(0, len(List), segment_size):
        yield List[i: i + segment_size]


def list_sliding_window(sequence, window_size=2):
    """ Returns a sliding window generator (of size window_size) over a sequence. Taken from
        https://docs.python.org/release/2.3.5/lib/itertools-example.html

        Example
        -------
        List = ['a', 'b', 'c', 'd']
        window_size = 2
        list_sliding_window(List, 2) -> ('a', 'b')
                                        ('b', 'c')
                                        ('c', 'd')

        Parameters
        ----------
        sequence: sequence (string, unicode, list, tuple, etc.)
            Sequence to be iterated over. window_size=1 is just a regular iterator.
        window_size: int
            Size of each window.

        Returns
        -------
        Generator

    """
    iterable = iter(sequence)
    result = tuple(islice(iterable, window_size))
    if len(result) == window_size:
        yield result
    for item in iterable:
        result = result[1:] + (item,)
        yield result


class LexicalRichness(object):
    """ Object containing tokenized text and methods to compute Lexical Richness (also known as
        Lexical Diversity or Vocabulary Diversity.)
    """

    def __init__(self, text, preprocessor=preprocess, tokenizer=tokenize):
        """ Initialise object with basic attributes needed to compute the common lexical diversity measures.

            Parameters
            ----------
            text: string or list
                String (or unicode) variable containing textual data, or a list
                of tokens if the text is already tokenized.
            preprocessor: callable or None
                A callable for preprocessing the text. Default is the built-in
                `preprocess` function. If None, no preprocessing is applied.
            tokenizer: callable or None
                A callable for tokenizing the text. Default is the built-in
                `tokenize` function. If None, the text parameter should be a list.

            Attributes
            ----------
            wordlist: list
                List of tokens from text.
            words: int
                Number of words in text.
            terms: int
                Number of unique terms/vocabb in text.
            preprocessor: callable
                The preprocessor used.
            tokenizer: callable
                The tokenizer used.

            Helpers Functions
            -----------------
            preprocess(string):
                Preprocess text before tokenizing (if preprocessor=preprocess)
            blobber(string)
                Tokenize text using TextBlob (if tokenizer=blobber)
            tokenize(string)
                Tokenize text using built-in string methods (if tokenizer=tokenize)
        """
        self.preprocessor = preprocessor
        self.tokenizer = tokenizer

        if self.preprocessor:
            text = self.preprocessor(text)

        if self.tokenizer:
            self.wordlist = self.tokenizer(text)
        else:
            self.wordlist = text

        self.words = len(self.wordlist)
        self.terms = len(set(self.wordlist))


    # Lexical richness measures
    @property
    def ttr(self):
        """ Type-token ratio (TTR) computed as t/w, where t is the number of unique terms/vocab,
            and w is the total number of words.
            (Chotlos 1944, Templin 1957)
        """
        return self.terms / self.words


    @property
    def rttr(self):
        """ Root TTR (RTTR) computed as t/sqrt(w), where t is the number of unique terms/vocab,
            and w is the total number of words.
            Also known as Guiraud's R and Guiraud's index.
            (Guiraud 1954, 1960)
        """
        return self.terms / sqrt(self.words)


    @property
    def cttr(self):
        """ Corrected TTR (CTTR) computed as t/sqrt(2 * w), where t is the number of unique terms/vocab,
            and w is the total number of words.
            (Carrol 1964)
        """
        return self.terms / sqrt(2 * self.words)


    @property
    def Herdan(self):
        """ Computed as log(t)/log(w), where t is the number of unique terms/vocab, and w is the
            total number of words.
            Also known as Herdan's C.
            (Herdan 1960, 1964)
        """
        return log(self.terms) / log(self.words)


    @property
    def Summer(self):
        """ Computed as log(log(t)) / log(log(w)), where t is the number of unique terms/vocab, and
            w is the total number of words.
            (Summer 1966)
        """
        return log(log(self.terms)) / log(log(self.words))


    @property
    def Dugast(self):
        """ Computed as (log(w) ** 2) / (log(w) - log(t)), where t is the number of unique terms/vocab,
            and w is the total number of words.
            (Dugast 1978)
        """
        # raise exception if terms and words count are the same
        if self.words == self.terms:
            raise ZeroDivisionError('Word count and term counts are the same.')

        return (log(self.words) ** 2) / (log(self.words) - log(self.terms))


    @property
    def Maas(self):
        """ Maas's TTR, computed as (log(w) - log(t)) / (log(w) * log(w)), where t is the number of
            unique terms/vocab, and w is the total number of words. Unlike the other measures, lower
            maas measure indicates higher lexical richness.
            (Maas 1972)
        """
        return (log(self.words) - log(self.terms)) / (log(self.words) ** 2)


    def msttr(self, segment_window=100, discard=True):
        """ Mean segmental TTR (MSTTR) computed as average of TTR scores for segments in a text.

            Split a text into segments of length segment_window. For each segment, compute the TTR.
            MSTTR score is the sum of these scores divided by the number of segments.
            (Johnson 1944)

            Helper Function
            ---------------
            segment_generator(List, segment_window):
                Split a list into s segments of size r (segment_size).

            Parameters
            ----------
            segment_window: int
                Size of each segment (default=100).
            discard: bool
                If True, discard the remaining segment (e.g. for a text size of 105 and a segment_window
                of 100, the last 5 tokens will be discarded). Default is True.

            Returns
            -------
            float
        """
        if segment_window >= self.words:
            raise ValueError('Window size must be greater than text size of {}. Try a smaller segment_window size.'.format(self.words))

        if segment_window < 1 or type(segment_window) is float:
            raise ValueError('Window size must be a positive integer.')

        scores = list()
        for segment in segment_generator(self.wordlist, segment_window):
            ttr = len(set(segment)) / len(segment)
            scores.append(ttr)

        if discard: # discard remaining words
            del scores[-1]

        if sys.version_info == 3:
            mean_ttr = mean(scores)
        else:
            mean_ttr = sum(scores) / len(scores)
        return mean_ttr


    def mattr(self, window_size=100):
        """ Moving average TTR (MATTR) computed using the average of TTRs over successive segments
            of a text.

            Estimate TTR for tokens 1 to n, 2 to n+1, 3 to n+2, and so on until the end
            of the text (where n is window size), then take the average.
            (Covington 2007, Covington and McFall 2010)

            Helper Function
            ---------------
            list_sliding_window(sequence, window_size):
                Returns a sliding window generator (of size window_size) over a sequence

            Parameter
            ---------
            window_size: int
                Size of each sliding window.

            Returns
            -------
            float
        """
        if window_size > self.words:
            raise ValueError('Window size must not be greater than text size of {}. Try a smaller window size.'.format(self.words))

        if window_size < 1 or type(window_size) is float:
            raise ValueError('Window size must be a positive integer.')

        scores = [len(set(window)) / window_size
                  for window in list_sliding_window(self.wordlist, window_size)]

        if sys.version_info == 3:
            mattr = mean(scores)
        else:
            mattr = sum(scores)/len(scores)

        return mattr


    def mtld(self, threshold=0.72):
        """ Measure of textual lexical diversity, computed as the mean length of sequential words in
            a text that maintains a minimum threshold TTR score.

            Iterates over words until TTR scores falls below a threshold, then increase factor
            counter by 1 and start over. McCarthy and Jarvis (2010, pg. 385) recommends a factor
            threshold in the range of [0.660, 0.750].
            (McCarthy 2005, McCarthy and Jarvis 2010)

            Parameters
            ----------
            threshold: float
                Factor threshold for MTLD. Algorithm skips to a new segment when TTR goes below the
                threshold (default=0.72).

            Returns
            -------
            float
        """

        def sub_mtld(self, threshold, reverse=False):
            """
            Parameters
            ----------
            threshold: float
                Factor threshold for MTLD. Algorithm skips to a new segment when TTR goes below the
                threshold (default=0.72).
            reverse: bool
                If True, compute mtld for the reversed sequence of text (default=False).

            Returns:
                mtld measure (float)
            """
            if reverse:
                word_iterator = iter(reversed(self.wordlist))
            else:
                word_iterator = iter(self.wordlist)

            terms = set()
            word_counter = 0
            factor_count = 0

            for word in word_iterator:
                word_counter += 1
                terms.add(word)
                ttr = len(terms)/word_counter

                if ttr <= threshold:
                    word_counter = 0
                    terms = set()
                    factor_count += 1

            # partial factors for the last segment computed as the ratio of how far away ttr is from
            # unit, to how far away threshold is to unit
            if word_counter > 0:
                factor_count += (1-ttr) / (1 - threshold)

            # ttr never drops below threshold by end of text
            if factor_count == 0:
                ttr = self.terms / self.words
                if ttr == 1:
                    factor_count += 1
                else:
                    factor_count += (1-ttr) / (1 - threshold)

            return len(self.wordlist) / factor_count

        forward_measure = sub_mtld(self, threshold, reverse=False)
        reverse_measure = sub_mtld(self, threshold, reverse=True)

        if sys.version_info[0] == 3:
            mtld = mean((forward_measure, reverse_measure))
        else:
            mtld = (forward_measure + reverse_measure) / 2

        return mtld


    def hdd(self, draws=42):
        """ Hypergeometric distribution diversity (HD-D) score.

            For each term (t) in the text, compute the probabiltiy (p) of getting at least one appearance
            of t with a random draw of size n < N (text size). The contribution of t to the final HD-D
            score is p * (1/n). The final HD-D score thus sums over p * (1/n) with p computed for
            each term t. Described in McCarthy and Javis 2007, p.g. 465-466.
            (McCarthy and Jarvis 2007)

            Parameters
            ----------
            draws: int
                Number of random draws in the hypergeometric distribution (default=42).

            Returns
            -------
            float
        """
        if self.terms < 42:
            suggestion = self.words // 2
        else:
            suggestion = 42
        if self.words < draws:
            raise ValueError('Number of draws should be less than the total sample size of {0}. Try a draw value smaller than {0}, e.g. hdd(draws={1}.)'.format(self.words, suggestion))
        if draws < 1 or type(draws) is float:
            raise ValueError('Number of draws must be a positive integer. E.g. hdd(draws={})'.format(suggestion))

        term_freq = Counter(self.wordlist)

        term_contributions = [(1 - hypergeom.pmf(0, self.words, freq, draws)) / draws
                              for term, freq in term_freq.items()]

        return sum(term_contributions)


    def __str__(self):
        return ' '.join(self.wordlist)


    def __repr__(self):
        return 'LexicalRichness(words={}, terms={}, preprocessor={}, tokenizer={}, wordlist={}, string="{}")'.format(
            self.words, self.terms, self.preprocessor, self.tokenizer, self.wordlist, ' '.join(self.wordlist)
        )
